src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -389,41 +315,210 @@ private static <T extends Enum<T>> int mappedEnumLookup(T value, MethodHandles.L
+     *     ...
+     * }
+     */
+    @SuppressWarnings("removal")
+    private static MethodHandle generateInnerClass(MethodHandles.Lookup caller, Object[] labels) {
+        List<EnumDesc<?>> enumDescs = new ArrayList<>();
+
+        byte[] classBytes = Classfile.of().build(ClassDesc.of(typeSwitchClassName(caller.lookupClass())), clb -> {
+            clb.withFlags(AccessFlag.FINAL, AccessFlag.SYNTHETIC)
+              .withMethod("typeSwitch", MethodTypeDesc.ofDescriptor("(Ljava/lang/Object;ILjava/util/function/BiFunction;)I"), Classfile.ACC_STATIC, mb -> {
+                  mb.withFlags(AccessFlag.PUBLIC, AccessFlag.FINAL, AccessFlag.STATIC)
+                    .withCode(cb -> {
+                        cb.aload(0);
+                        Label nonNullLabel = cb.newLabel();
+                        cb.if_nonnull(nonNullLabel);
+                        cb.iconst_m1();
+                        cb.ireturn();
+                        cb.labelBinding(nonNullLabel);
+                        if (labels.length == 0) {
+                            cb.constantInstruction(0)
+                              .ireturn();
+                            return ;
+                        }
+                        cb.iload(1);
+                        Label dflt = cb.newLabel();
+                        record Element(Label target, Label next, Object label) {}
+                        List<Element> cases = new ArrayList<>();
+                        List<SwitchCase> switchCases = new ArrayList<>();
+                        Object lastLabel = null;
+                        for (int idx = labels.length - 1; idx >= 0; idx--) {
+                            Object currentLabel = labels[idx];
+                            Label target = cb.newLabel();
+                            Label next;
+                            if (lastLabel == null) {
+                                next = dflt;
+                            } else if (lastLabel.equals(currentLabel)) {
+                                next = cases.getLast().next();
+                            } else {
+                                next = cases.getLast().target();
+                            }
+                            lastLabel = currentLabel;
+                            cases.add(new Element(target, next, currentLabel));
+                            switchCases.add(SwitchCase.of(idx, target));
+                        }
+                        cases = cases.reversed();
+                        switchCases = switchCases.reversed();
+                        cb.tableswitch(0, labels.length - 1, dflt, switchCases);
+                        for (int idx = 0; idx < cases.size(); idx++) {
+                            Element element = cases.get(idx);
+                            Label next = element.next();
+                            cb.labelBinding(element.target());
+                            if (element.label() instanceof Class<?> classLabel) {
```

Comment:
It's too bad we can not use a switch on the label here instead of a bunch of instanceof :)

Reply:
Well, we could, if we tweaked javac so that it would produce a different/simplified code for java.base for pattern switches (i.e. hardcode an if-else cascade). Which we may or may not do at some point.